design/one-pager-resource-connectivity-mvp.md

```diff
@@ -0,0 +1,653 @@
+   network.
+1. The app operator creates their `KubernetesCluster` and `MySQLInstance`
+   resource claims, which either explicitly reference or default to the resource
+   classes created in step 2. Because their underlying managed resources are
+   appropriately configured, they can now communicate!
+
+The remainder of this proposal works through this scenario in each cloud
+provider, highlighting the changes necessary to enable connectivity.
+
+### Resource References
+
+It's typical for Kubernetes resources to refer to each other. The [Kubernetes
+API conventions] state:
+
+> References to loosely coupled sets of objects, such as pods overseen by a
+> replication controller, are usually best referred to using a label selector.
+> [...] References to specific objects, especially specific resource versions
+> and/or specific fields of those objects, are specified using the
+> ObjectReference type (or other types representing strict subsets of it).
+> [...] Object references should either be called fooName if referring to an
+> object of kind Foo by just the name (within the current namespace, if a
+> namespaced resource), or should be called fooRef, and should contain a subset
+> of the fields of the ObjectReference type.
+
+Crossplane currently uses these conventions to create references between
+resource claims and classes, resource claims and managed resources, etc. No
+pattern currently exists for modelling relationships between managed resources.
+
+Assume a `Network` managed resource named `kubernetesname` exists in namespace
+`crossplane-system`. Further assume this managed resource represents a GCP VPC
+network whose 'real' name in the GCP API is `externalname`. Finally, assume a
+`CloudSQLInstance` managed resource wants to specify this network as its
+`.spec.ipConfiguration.privateNetwork` (i.e. whitelisted VPC network). This
+could be implemented in one of two ways:
+
+1. The `CloudSQLInstance` simply requires `.spec.ipConfiguration.privateNetwork`
+   be provided as a `string` name specifying the real, external name of the
+   `Network`, i.e. `externalname`. It is up to the creator of the
+   `CloudSQLInstance` (or the `ResourceClass` used to create it) to ensure this
+   name corresponds with the name of the `kubernetesname` `Network`.
+1. The `CloudSQLInstance` requires an `ObjectReference` at be specified as
+   `.spec.ipConfiguration.privateNetworkRef`. This `ObjectReference` refers to
+   `kind: Network`, `namespace: crossplane-system`, `name: kubernetesname`. The
+   `CloudSQLInstance` controller must then lookup the specified `Network` in
+   order to determine that it is named `externalname` when submitting requests
+   to the GCP API.
+
+This document suggests the first approach, both because it requires less logic
+and fewer Kubernetes API calls to implement, and because it enables Crossplane
+users to reference both external resources that are and are not modelled as
+Crossplane managed resources.
```

Comment:
The first time I read this I thought I disagreed that the first approach was a good idea. However, when thinking about it more, I agree that the first approach provides the most flexibility and enables a Crossplane user to manage different components in different places. For instance, I could see a user saying they want to set up all of the network resources on the provider platform directly, but handle all other resources in Crossplane. I think this represents a general outlook that we should probably have for all aspects of Crossplane: allow more, dictate less. We can always enable more rigid rules with additional tooling and layers, but it is much harder to work around strict decisions that have been implemented as part of the platform.

Reply:
My thoughts on this topic followed a similar pattern. I initially preferred object references, however:

* If we use object references, we can only model relationships with external resources that have corresponding Crossplane managed resources. Presuming we wanted to support folks bringing their existing external resources, we'd need some equivalent to Terraform's `import` and `state rm` functionality, to attach and detach external resources from Crossplane management. We'd also likely need some equivalent to Terraform's data sources, i.e. read-only representations of external resources.
* Using object references at the managed resource level _slightly_ blurs our ability to write high fidelity CRDs in which the fields map unsurprisingly to the cloud provider API. For example the `CloudSQLInstance` `privateNetwork` field referenced elsewhere in this document becomes `privateNetworkRef`.

However, there are also shortcomings with referencing objects by their external name, chiefly among which is that we frequently automatically generate the external name and thus it may not be known until after the managed resource is created.